### DISCUSSION TITLE
Help to use S3 on controller

### DISCUSSION THREAD

**Discussion by brandonsousa** [adonisjs/core#1383] [Help] [answered]
I have a controller that when registered the user I have to upload his photo in aws s3, all the data comes with the form, to save the local photo it works, but to send to s3 it doesn't send, i already tried to use the bodyparser .processManually and request.multipart.file but it doesn't suit my UC, I'm also trying this way from the code below which is what doesn't upload too

`
const newAdv = request.except('img_url')
const image = request.file('img_url', {
  types: ['png', 'jpg', 'jpeg'],
  size: '4mb'
})
if (cpf.isValid(newAdv.document) || cnpj.isValid(newAdv.document)) {
  const a = await Advertiser.create({ 'user_id': auth.user.id, ...newAdv })
  if (a) {

    const path = '/assets/images/media/'
    
    const nameImage = `${new Date().getTime()}.${image.subtype}`

    await Drive.disk('s3').put(nameImage, image.stream)

    await image.move(Helpers.publicPath(path), {
      name: nameImage
    })

    a.merge({
      'img_url': 'http://aki-perto.com/' + path + nameImage
    })

    await a.save()
    session.flash({ success: 'Cadastrado com sucesso' })
    return response.redirect('/advertiser/create')
  }
  session.flash({ error: 'Erro ao cadastrar' })
  return response.redirect('/advertiser/create')
}
session.flash({ error: 'Erro ao cadastrar, documento inválido' })
return response.redirect('/advertiser/create')`

**Comment by McSneaky** [comment]
Do you get any error or anything?
What's your config currently, processManually enabled or disabled for given route?
Here's an example how to do it: 
https://adonisjs.com/docs/4.1/file-uploads#_process_the_stream

```js
const Drive = use('Drive')

Route.post('upload', async ({ request }) => {

  request.multipart.file('profile_pic', {}, async (file) => {
    await Drive.disk('s3').put(file.clientName, file.stream)
  })

  await request.multipart.process()
})
```

> You must call await request.multipart.process() to start processing uploaded files.

**Reply by brandonsousa** [reply]
the error returned is: 
`E_CANNOT_PROCESS_STREAM: Cannot process multipart stream twice. Make sure to disable files {autoProcess} when manually calling multipart.process > More details: https://err.sh/adonisjs/errors/E_CANNOT_PROCESS_STREAM`

**my bodyparser:**
`autoProcess: true,`
`processManually:` []`

i have a view on resources/views/create.edge
`<form action="/advertiser" method="POST" enctype="multipart/form-data">...`

i send data to my controller with image and other data

i  had seen it in some places that created a function just to send the image but it doesn't fit in this UC

**Reply by faiyaz26** [reply]
You need to do this part: https://adonisjs.com/docs/4.1/file-uploads#_disable_auto_processing

**Reply by brandonsousa** [reply]
I already used it, but my code doesn't run completely, if I use a separate controller just to upload and create a route for it, it works, but if I do it the way I'm trying, in case I create the user, do the upload after creation, everything inside the same controller it does not work, if I disable auto processing for the route and it is to register the code does not run, I was wondering if you can do everything together or I will have to create one separate route to upload

**Comment by faiyaz26** [comment, marked as answer]
to understand your use-case properly. You want to use same route for both uploading image and some form data. right?

If you want to use same route for uploading image along with some form field you should do the following:

1. disable autoprocessing. https://adonisjs.com/docs/4.1/file-uploads#_disable_auto_processing
2. follow this method to get the stream and pass the stream to S3. https://adonisjs.com/docs/4.1/file-uploads#_process_the_stream
3. Now if you use processManually, adonis will not be able to parse the information (AFAIK), so to get the fields data you need to do the following:
```
const fields = new Map(); // it will contain all the fields value
await request.multipart.field((name, value) => {
     fields.set(name, value);
    })
```
4. Upload the image using following code:
```
const Drive = use('Drive')

Route.post('upload', async ({ request }) => {

  request.multipart.file('profile_pic', {}, async (file) => {
    await Drive.disk('s3').put(file.clientName, file.stream)
  })

  await request.multipart.process()
})
```

Let me know if I have misunderstood your use-case.

**Reply by brandonsousa** [reply]
that's right, i want to use the same route to save the data and the photo in s3, as shown above my method for registration

**Reply by faiyaz26** [reply]
can you let me know whether the above solution works for you?

**Reply by brandonsousa** [reply]
It didn't work, I did it, I disabled autoprocessing and tried, but it fell right into the error, without executing all the code

`const newAdv = request.except('img_url')
const image = request.file('img_url', {
types: ['png', 'jpg', 'jpeg'],
size: '4mb'
})
if (cpf.isValid(newAdv.document) || cnpj.isValid(newAdv.document)) {
const a = await Advertiser.create({ 'user_id': auth.user.id, ...newAdv })
if (a) {

const path = '/assets/images/media/'

const nameImage = `${new Date().getTime()}.${image.subtype}`

await Drive.disk('s3').put(nameImage, image.stream)

await image.move(Helpers.publicPath(path), {
  name: nameImage
})

a.merge({
  'img_url': 'http://aki-perto.com/' + path + nameImage
})

await a.save()
session.flash({ success: 'Cadastrado com sucesso' })
return response.redirect('/advertiser/create')
}
session.flash({ error: 'Erro ao cadastrar' })
return response.redirect('/advertiser/create')
}
session.flash({ error: 'Erro ao cadastrar, documento inválido' })
return response.redirect('/advertiser/create')`

The part of the code he executed:
`session.flash({ error: 'Erro ao cadastrar, documento inválido' })
return response.redirect('/advertiser/create')`

**Reply by faiyaz26** [reply]
can you please check the code provided above? your code doesn't reflect the code I provided.

**Reply by brandonsousa** [reply]
it worked, I got it, but when I did it that way for the insomnia I got it, but by my form that is on the route "advertiser / create" that sends to the route that I created "register / advertiser" it sends it to the route "advertiser / register / create "

**Form**
`<form action="register/advertiser" method="POST" enctype="multipart/form-data">...</form>`

**Routes**
`Route.resource('advertiser', 'AdvertiserController').only(['index', 'create', 'show', 'update', 'destroy']).middleware('auth'))`
`Route.group(() => {
    Route.post('advertiser', 'AdvertiserController.store')
}).prefix('register').middleware('auth')`

**Comment by McSneaky** [comment]
Quite confused, what's the problem you are having. But here's potential solution if I understood it correctly:

Add `/` at the start of form action. Otherwise it just appends `register/advertiser` to current URL
Make it like this `action="/register/advertiser"`, `/` indicates to not append it, but to add it to domain root

**Reply by brandonsousa** [reply]
thank you, it worked, I had this problem for some time, but I left it aside, now that I had to go up to production there was no way to escape, thank you very much
